File: C++/958. Check Completeness of a Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    queue<TreeNode*> my_queue;
    vector<int> vec;
    void level_traverse(TreeNode* root) {
        my_queue.push(root);
        while (!my_queue.empty()) {
            TreeNode* temp = my_queue.front();
            vec.push_back(temp->val);
            my_queue.pop();
            if (temp->left)
                my_queue.push(temp->left);
            else if (temp->val != -1)
                my_queue.push(new TreeNode(-1));
            if (temp->right)
                my_queue.push(temp->right);
            else if (temp->val != -1)
                my_queue.push(new TreeNode(-1));
        }
    }
    bool isCompleteTree(TreeNode* root) {
        //Hierarchical traversal using queues
        level_traverse(root);
        //Null (ie -1) appears in the middle of the hierarchical traversal result, not a complete binary tree
        int flag_1 = 0, flag_2 = 0;
        for (int i = 0; i < vec.size(); i++) {
            if (vec[i] == -1)
                flag_1 = 1;
            if (flag_1 == 1 && vec[i] != -1)
                flag_2 = 1;
        }
        if (flag_1 == 1 && flag_2 == 1)
            return false;
        else
            return true;
    }
};

File: Python/958. Check Completeness of a Binary Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution:

    def level_traverse(root, level_order, index):
        if root and index < 300:
            level_order[index] = root.val
        if root.left:
            Solution.level_traverse(root.left, level_order, index*2)
        if root.right:
            Solution.level_traverse(root.right, level_order, index*2+1)

    def isCompleteTree(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        # The tree will have between 1 and 100 nodes.  => complete binary tree max_index < 300
        level_order = [-1 for i in range(300)]
        # Recursively build hierarchical traversal
        Solution.level_traverse(root, level_order, 1)

        # Null (ie -1) appears in the middle of the hierarchical traversal result, not a complete binary tree
        flag = 0
        for i in range(1, 300):
            if level_order[i] == -1:
                flag = 1
            if flag == 1 and level_order[i] != -1:
                flag = 2
                break

        if flag == 2:
            return False
        else:
            return True
